File: 2023/kotlin/src/Day09.kt
private fun parseReport(input: String) =
    input.lines().map { it.ints() }

private fun extrapolate(history: List<Int>) =
    generateSequence(history) { seq ->
        seq.zipWithNext { a, b -> b - a }.takeIf { it.any { diff -> diff != 0 } }
    }.fold(0) {
        acc, seq -> acc + seq.last()
    }

private fun part1(input: String) =
    parseReport(input).sumOf(::extrapolate)

private fun part2(input: String) =
    parseReport(input).map { it.reversed() }.sumOf(::extrapolate)

fun main() {
    val input = readInput("Day09")
    part1(input).println()
    part2(input).println()
}

File: 2023/swift/AdventOfCode2023/Sources/Day09.swift
import Algorithms

struct Day09: AdventDay {
  var input: String

  func parseReport(_ input: String) -> [[Int]] {
    input.lines().map { $0.ints() }
  }

  func extrapolate(_ history: [Int]) -> Int {
    if history.allSatisfy({ $0 == 0 }) {
      return 0
    }

    let diffs = history.windows(ofCount: 2).map { pair in pair.last! - pair.first! }

    return history.last! + extrapolate(diffs)
  }

  func part1() -> Any {
    parseReport(input)
      .map(extrapolate)
      .sum()
  }

  func part2() -> Any {
    parseReport(input)
      .map { $0.reversed() }
      .map(extrapolate)
      .sum()
  }
}
